### ISSUE TITLE
Averaging consumption rates of devices that does not have consistent power draw

### ISSUE THREAD

**Issue by Samahu** [nasa/GSAP#148]
Follow up on #147 and since we need to provide an instantaneous measurement of the power draw. Assuming we have a device that doesn't have a consistent power consumption during its operation. i.e. the device power draw fluctuates between 0.1 to 0.8 watts for example. So if we are sampling at low rate of 2 seconds or even less than that. We can get very different values sent to the GSAP progonser depending on when sampling takes places. For this case, does it make sense to maintain a moving average of the device consumption and send that as a measurement of the device draw rate at any given timestamp?

**Comment by teubert** [comment]
Hi @Samahu,

In general, my take is that if you're doing a state estimation step each time there's new data the Kalman or Particle Filter (see Observers in this package) to account for the noise. I generally advise a workflow like this. 

1. Every time there's new data- run a state estimator step
2. At fixed frequency or when last prediction ended- start a new prediction

You can do this by splitting out the observer and predictor (rather than using the combined prognoser). I'd be happy to write up an example that shows you how to do this if you'd like. 

However, from what I understand of your problem, the power is coming at a much higher rate than the rest of the data, and it contains a lot of fluctuation at very fast frequencies- applying a moving average to only the power draw before being fed into the state estimator seems like a fair approach

**Comment by Samahu** [comment]
Yes, we have a device that can provide power draw measurements at 50 Hz. However, the GSAP prognoser can take over a second to finish. That being there is no way we can implement the workflow /1/ that you mentioned. Our current code in fact implements the second workflow. However, since we are only performing the prediction step every two seconds that means we have over a 100 sample of power usage to pick from. The device power draw is inconsistent and without the moving average the power draw values may seem random to the prognoser.

> You can do this by splitting out the observer and predictor (rather than using the combined prognoser). I'd be happy to write up an example that shows you how to do this if you'd like.

If you think that would be a better approach that using the moving average then by all means.

**Comment by teubert** [comment]
Hi @Samahu,

I'm not sure we're on the same page. Both of my proposals involved using the state estimator/observer. 

Just to clarify- when you say "The GSAP prognoser can take over a second to finish" - that's both the state estimation (observer) and prediction step. Almost all of that time (I think I benchmarked it at between 99.5-99.9%) is in the predict step. My proposal 1 was to break out the state estimation and prediction steps so you can use the state estimation steps in GSAP instead or with the moving average. It shouldn't be much slower than the moving average (unless you use a Particle Filter with A LOT of samples). 

A moving average may be appropriate for some cases, but beware: in cases where the power draw has more than just random noise, moving average will lag behind the true state. In those cases I recommend using the included observers/state estimators 

To summarize clarification:
* Prognoser.step() = observer.step() + predictor.step() <- this is what you observed at >1 second
* My proposal 1 - break apart observer and predictor (common use case. It'd be a pretty good idea to have an example anyway)
  * run observer each time you have new power draw measurements 
  * Run predictor step every 3 seconds with the latest state from the observer.

**Comment by Samahu** [comment]
In my last comment I asked if you can provide an example for splitting the observer and the predictor, I am not clear on how to split the prognoser into separate observer and predictor objects based on the provided library example: https://github.com/nasa/GSAP/blob/ea3d4ad1ce5cef2c0e7c1d0d087b0b0aaa142acf/examples/simple/main.cpp#L80

**Comment by Samahu** [comment]
Just to note I think you are referring to something similar to this:
https://github.com/nasa/GSAP/blob/ea3d4ad1ce5cef2c0e7c1d0d087b0b0aaa142acf/src/ModelBasedPrognoser.cpp#L154-L168

But I feel that these objects are internal to the library and are not meant for end users.
